### PR TITLE
Implementation of the Zhang, Shu 2D Positivity Preserving Limiter - Generalized for all dimensions

### PR DESCRIPTION
**Pull request by ssaishruthi98**
Modified the Positivity Preserving Limiter (PPL) implementation to include mixed nodes as per the Nov 2010 Zhang Shu Paper and generalized it for all dimensions. 

Implemented several new test cases for validation such as the Double Mach Reflection, Astrophysical Jet, Strong Vortex Shock Wave and Shock Diffraction case. Added a few new boundary conditions for these cases as well including a custom boundary, p0 outflow boundary and an inflow boundary specific to the Astrophysical Jet case. The pre-existing Low Density test case has been modified to be a Gaussian Pulse test case instead of the sine wave as it is a better test for positivity-preservation.

Moved all the test cases related to the PPL out of the non_periodic_cube_flow flow solver case and made a separate flow solver case called positivity_preserving_tests. Grid file of the same name has been included with a grid function for each case that's been added. New function has also been added within the new flow solver case for calculating entropy. 

Main changes of note are within the positivity_preserving_limiter.cpp file.

Edit: Changes were made to the Maximum Principle Limiter to use the mixed nodes as well but only for the purpose of finding the maximum and minimum values in each cell. The implementation of the mixed node solution cell average for this limiter hasn't been done, so functionally nothing should have changed. I reran the OOA Burger's test to verify that the orders are still the same and nothing has changed from the mixed node implementation.

### CODE REVIEW

File: src/limiter/maximum_principle_limiter.cpp [hunk 1]
```diff
@@ -207,17 +238,11 @@ void MaximumPrincipleLimiter<dim, nstate, real>::limit(
         // Apply limiter on solution values at quadrature points
         for (unsigned int istate = 0; istate < nstate; ++istate) {
             for (unsigned int iquad = 0; iquad < n_quad_pts; ++iquad) {
-                soln_at_q[istate][iquad] = theta[istate] * (soln_at_q[istate][iquad] - soln_cell_avg[istate])
+                soln_coeff[istate][iquad] = theta[istate] * (soln_coeff[istate][iquad] - soln_cell_avg[istate])
```
Comment (sivanadarajah) [inline src/limiter/maximum_principle_limiter.cpp:241]: This is just a placeholder that we should return and show mathematically whether updating the solution coefficients ensures we can preserve the maximum principle instead of the solution at the quadrature nodes.

File: src/limiter/positivity_preserving_limiter.cpp [hunk 2]
```diff
@@ -172,8 +187,160 @@ void PositivityPreservingLimiter<dim, nstate, real>::write_limited_solution(
...
+                    for(unsigned int jquad=0; jquad<quad_weights_GLL.size(); ++jquad) {
+                        soln_cell_avg_dim[1][istate] += quad_weights_GL[iquad]*quad_weights_GLL[jquad]*soln_at_q[1][istate][quad_pt];
+                            quad_pt++;
+                    }
+                }
+            }
+        }
+
+        if (dim == 3) {
+            for(unsigned int istate = 0; istate < nstate; ++istate) {
```
Comment (sivanadarajah) [inline src/limiter/positivity_preserving_limiter.cpp:245]: For the 3D case do we need to consider six options or does their paper just limit it to GLL x GL x GL? If they did not it would be interesting to see how the analysis/math turns out.

**Review by sivanadarajah — COMMENTED** [review]
Nice work on this PR. I just went through the limiters for now.

**Comment by jbrillon** [comment]
why did this get closed?

**Review by jbrillon — CHANGES_REQUESTED** [review]
as discussed, create a branch off this branch called `2DLimiterImplementation_CleanUp` and push all improvements to it, then once it's ready, merge `2DLimiterImplementation_CleanUp` into `2DLimiterImplementation`

**Review by cpethrick — COMMENTED** [review]
Thanks for this PR, having these updates to the limiters in the main branch will make it even easier for everyone to run complex cases. My comments are mostly minor and I hope they're not a lot of work to resolve. Great job on this!

I have one general comment about the implementation. How costly is it to check all node combinations, and is it required in all cases? The first part of that question could be answered by running a smooth case with the old and new versions of a limiter (not sure whether MPL or PPL would be more suitable there), while the second part of the question can probably be addressed through your experience running test cases so far.

File: src/physics/euler.cpp [hunk 3]
```diff
@@ -1319,6 +1323,61 @@ ::boundary_farfield (
...
+
+    const std::array<real, nstate> conservative_bc = convert_primitive_to_conservative(primitive_boundary_values);
+    for (int istate = 0; istate < nstate; ++istate) {
+        soln_bc[istate] = conservative_bc[istate];
+    }
+}
+
+template <int dim, int nstate, typename real>
+void Euler<dim, nstate, real>
+::boundary_astrophysical_inflow(
```
[the remark this answers is not in the document]
Comment (ssaishruthi98) [reply src/physics/euler.cpp:1354]: The astrophysical jet case unfortunately requires two custom boundaries. One to set the inflow of the jet itself from -0.05<y<0.05,x=0 and one to set the inflow for the rest of the boundary along x=0. 

Setting it up this way allows for the custom boundary to be used to change the strength of the jet inflow from Mach 80 to Mach 2000 and the astrophysical inflow boundary remains the same between both cases.

File: src/testing/bound_preserving_limiter_tests.cpp [hunk 4]
```diff
@@ -209,16 +240,27 @@ int BoundPreservingLimiterTests<dim, nstate>::run_convergence_test() const
...
+        convergence_table.set_scientific("soln_Linf_error", true);
         if (this->pcout.is_active()) convergence_table.write_text(this->pcout.get_stream());
+
+        std::ofstream table_file("convergence_rates.txt");
+        convergence_table.write_text(table_file);
+
+        
     }//end of grid loop
+
     return 0;
```
[the remark this answers is not in the document]
Comment (ssaishruthi98) [reply src/testing/bound_preserving_limiter_tests.cpp:258]: Unfortunately there is not easy pass/fail condition to add. Due to the nature of the Gaussian Pulse tests, the final orders don't always reach p+1 exactly but tend to be in the range of it. 

There are conditions being checked in the limiter convergence test flow solver cases for the principles that the limiters are supposed to conserve and if those are violated, then the test will fail there.

File: src/limiter/positivity_preserving_limiter.h [hunk 5]
```diff
@@ -80,6 +97,10 @@ class PositivityPreservingLimiter : public BoundPreservingLimiterState <dim, nst
         const unsigned int                                      n_shape_fns,
         const std::vector<dealii::types::global_dof_index>&     current_dofs_indices);
 
+    // Values required to compute solution cell average in 2D/3D
+    real dx;
```
[the remark this answers is not in the document]
Comment (ssaishruthi98) [reply src/limiter/positivity_preserving_limiter.h]: Added documentation!

File: src/limiter/positivity_preserving_limiter.cpp [hunk 6]
```diff
@@ -172,8 +184,156 @@ void PositivityPreservingLimiter<dim, nstate, real>::write_limited_solution(
...
+                    }
+                }
+            }
+        }
+
+        for (unsigned int istate = 0; istate < nstate; ++istate) {
+            soln_cell_avg[istate] = 0;
+        }
+
+        real lambda_1 = dt/this->dx; real lambda_2 = dt/this->dy; real lambda_3 = 0.0;
```
[the remark this answers is not in the document]
Comment (ssaishruthi98) [reply src/limiter/positivity_preserving_limiter.cpp]: Lambda is what was used to represent these values in eqn3.8 of the Zhang and Shu paper. I've left a comment indicating that in the code.

File: src/mesh/grids/positivity_preserving_tests_grid.cpp
[the remark this answers is not in the document]
Comment (ssaishruthi98) [reply src/mesh/grids/positivity_preserving_tests_grid.cpp:1]: I took a look through and added const where I could but there wasn't a lot from what I could tell. Please let me know if I missed any! 😄

File: src/operators/operators.cpp [hunk 7]
```diff
@@ -1512,7 +1512,7 @@ void local_Flux_Reconstruction_operator<dim,n_faces,real>::get_FR_correction_par
     }
     else if(FR_param_type == FR_enum::c10Thousand){ 
         //Set the value to 10000 for arbitrary high-numbers.
-        c = 10000.0;
+        c = 3.67e-3 * 10;//10000.0;
```
[the remark this answers is not in the document]
Comment (ssaishruthi98) [reply src/operators/operators.cpp]: Sorry, this was for testing a higher value of c for the astrophysical jet case. It has been reverted now.

File: src/parameters/parameters_flow_solver.cpp [hunk 8]
```diff
@@ -340,12 +389,17 @@ void FlowSolverParam::parse_parameters(dealii::ParameterHandler &prm)
...
-        else if (flow_case_type_string == "low_density_2d")             {flow_case_type = low_density_2d;}
+        else if (flow_case_type_string == "low_density")                {flow_case_type = low_density;}
         else if (flow_case_type_string == "leblanc_shock_tube")         {flow_case_type = leblanc_shock_tube;}
         else if (flow_case_type_string == "shu_osher_problem")          {flow_case_type = shu_osher_problem;}
         else if (flow_case_type_string == "advection_limiter")          {flow_case_type = advection_limiter;}
         else if (flow_case_type_string == "burgers_limiter")            {flow_case_type = burgers_limiter;}
+        else if (flow_case_type_string == "double_mach_reflection")     {flow_case_type = double_mach_reflection;}
+        else if (flow_case_type_string == "shock_diffraction")          {flow_case_type = shock_diffraction;}
+        else if (flow_case_type_string == "astrophysical_jet")          {flow_case_type = astrophysical_jet;}
+        else if (flow_case_type_string == "svsw")                       {flow_case_type = svsw;}
```
[the remark this answers is not in the document]
Comment (ssaishruthi98) [reply src/parameters/parameters_flow_solver.cpp]: Changed the parameter to strong_vortex_shock_wave. Left the grid functions and IC functions as SVSW. Is that ok?

File: src/physics/initial_conditions/initial_condition_function.cpp [hunk 9]
```diff
@@ -484,113 +525,220 @@ ::primitive_value(const dealii::Point<dim, real>& point, const unsigned int ista
...
         }
     }
     return value;
 }
 
 // ========================================================
-// 1D Leblanc Shock tube -- Initial Condition
-// See Zhang & Shu, On positivity-preserving..., 2010 Pg. 14
+// Double Mach Reflection Problem (2D) -- Initial Condition
+// INCLUDE REFERENCE LATER
```
[the remark this answers is not in the document]
Comment (ssaishruthi98) [reply src/physics/initial_conditions/initial_condition_function.cpp]: Included! 😅

File: src/flow_solver/flow_solver_cases/positivity_preserving_tests.h [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+class PositivityPreservingTests : public CubeFlow_UniformGrid<dim, nstate>
+{
+#if PHILIP_DIM==1
+     using Triangulation = dealii::Triangulation<PHILIP_DIM>;
+ #else
+     using Triangulation = dealii::parallel::distributed::Triangulation<PHILIP_DIM>;
+ #endif
+
+ public:
+     explicit PositivityPreservingTests(const Parameters::AllParameters *const parameters_input);
```
[the remark this answers is not in the document]
Comment (ssaishruthi98) [reply src/flow_solver/flow_solver_cases/positivity_preserving_tests.h]: Added documentation.

File: src/testing/bound_preserving_limiter_tests.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (cpethrick) [reply src/testing/bound_preserving_limiter_tests.cpp:258]: Can you check that it reaches the same OOA? e.g., if your version of the code converges at 3.7, check whether another version of the code will converge at/near 3.7

File: src/physics/euler.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (cpethrick) [reply src/physics/euler.cpp:1354]: Sounds good, thanks for clarifying that. Seems like it works well as an additional BC.

File: src/physics/initial_conditions/initial_condition_function.cpp
[the remark this answers is not in the document]
Comment (cpethrick) [reply src/physics/initial_conditions/initial_condition_function.cpp:1]: Looks good! Thanks.

File: src/parameters/parameters_flow_solver.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (cpethrick) [reply src/parameters/parameters_flow_solver.cpp]: That name is better. The `flow_case_string` appears everywhere that those grid/IC functions are found, right? If I've described it well, then the full name will appear.

File: src/operators/operators.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (cpethrick) [reply src/operators/operators.cpp]: Sweet, thanks!

File: src/mesh/grids/positivity_preserving_tests_grid.cpp
[the remark this answers is not in the document]
Comment (cpethrick) [reply src/mesh/grids/positivity_preserving_tests_grid.cpp:1]: Looks good.

File: src/limiter/positivity_preserving_limiter.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (cpethrick) [reply src/limiter/positivity_preserving_limiter.cpp]: Okay. My preference would usually be to use a non-symbolic name, but this is okay if you're directly referring to an equation in a paper.

File: src/testing/bound_preserving_limiter_tests.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (ssaishruthi98) [reply src/testing/bound_preserving_limiter_tests.cpp:258]: I could do this but since this test file is used by the 1D/2D Advection, 1D/2D Burgers and 1D/2D Low Density case, there would essentially be a lot of new repetitive code, checking for the order of each of those test cases. Some of those cases also have individual runs of Strong DG, NSFR without limiter and NSFR with limiter so there would have to be a distinction between each of these runs as well. Would this be ok?

[the remark this answers is not in the document]
Comment (ssaishruthi98) [reply src/testing/bound_preserving_limiter_tests.cpp:258]: Based on my conversation with @cpethrick, I'll set up a parameter for expected order for these cases so it can be passed in through the prm file instead of having a lot of new repeated code.

[the remark this answers is not in the document]
Comment (ssaishruthi98) [reply src/testing/bound_preserving_limiter_tests.cpp]: Added the parameter and included the parameter in the prm file for all the limiter convergence tests.

File: src/parameters/parameters_flow_solver.cpp [hunk 11]
```diff
@@ -284,6 +284,14 @@ void FlowSolverParam::declare_parameters(dealii::ParameterHandler &prm)
 
         }
         prm.leave_subsection();
+        
+        prm.enter_subsection("limiter_convergence_tests");
```
Comment (cpethrick) [inline src/parameters/parameters_flow_solver.cpp:288]: I would even prefer this not in its own subsection - I can see it being useful for h-refinement or time refinement tests. Thanks for adding it!

**Review by cpethrick — APPROVED** [review]
Leaving my approval - there are a few comments still unresolved but my concerns have been addressed. Great work!

File: src/parameters/parameters_flow_solver.cpp (hunk 11 above)
Comment (ssaishruthi98) [reply src/parameters/parameters_flow_solver.cpp]: Made it a general flow solver parameter! :smile:

File: src/limiter/tvb_limiter.h [hunk 12]
```diff
@@ -72,7 +72,8 @@ class TVBLimiter : public BoundPreservingLimiterState <dim, nstate, real>
         const unsigned int                                      grid_degree,
         const unsigned int                                      max_degree,
         const dealii::hp::FECollection<1>                       oneD_fe_collection_1state,
-        const dealii::hp::QCollection<1>                        oneD_quadrature_collection);
+        const dealii::hp::QCollection<1>                        oneD_quadrature_collection,
+        double                                                  dt);
```
[the remark this answers is not in the document]
Comment (ssaishruthi98) [reply src/limiter/tvb_limiter.h]: Added override for all the limit functions.

**Review by DominicRoy2 — APPROVED** [review]
Great work!

**Comment by ssaishruthi98** [comment]
> > Ran ctest on the branch and there were no unexpected failures. 😄 [LastTestsFailed.log](https://github.com/user-attachments/files/21582118/LastTestsFailed.log)
> 
> How much time is added for the new tests?

This PR adds 13 tests to ctest and increases run time by 15 mins.

File: src/testing/TGV_scaling.cpp [hunk 13]
```diff
@@ -44,8 +44,8 @@ int EulerTaylorGreenScaling<dim, nstate>::run_test() const
 
     std::ofstream myfile (all_parameters_new.energy_file + ".gpl"  , std::ios::trunc);
     const unsigned int poly_degree_start= all_parameters->flow_solver_param.poly_degree;
-
```
Comment (jbrillon) [inline src/testing/TGV_scaling.cpp:47]: avoid committing stuff like this that doesnt change the code, it can lead to more merge conflicts down the line

File: src/flow_solver/flow_solver_cases/cube_flow_uniform_grid.cpp [hunk 14]
```diff
@@ -18,7 +18,11 @@ double CubeFlow_UniformGrid<dim,nstate>::get_adaptive_time_step(std::shared_ptr<
 {
     // compute time step based on advection speed (i.e. maximum local wave speed)
     const unsigned int number_of_degrees_of_freedom_per_state = dg->dof_handler.n_dofs()/nstate;
-    const double approximate_grid_spacing = (this->all_param.flow_solver_param.grid_right_bound-this->all_param.flow_solver_param.grid_left_bound)/pow(number_of_degrees_of_freedom_per_state,(1.0/dim));
+    double approximate_grid_spacing = 0.0;
+    if(this->all_param.flow_solver_param.grid_xmax == this->all_param.flow_solver_param.grid_xmin)
+        approximate_grid_spacing = (this->all_param.flow_solver_param.grid_right_bound-this->all_param.flow_solver_param.grid_left_bound)/pow(number_of_degrees_of_freedom_per_state,(1.0/dim));
+    else
+        approximate_grid_spacing = (this->all_param.flow_solver_param.grid_xmax-this->all_param.flow_solver_param.grid_xmin)/pow(number_of_degrees_of_freedom_per_state,(1.0/dim));
```
[the remark this answers is not in the document]
Comment (ssaishruthi98) [reply src/flow_solver/flow_solver_cases/cube_flow_uniform_grid.cpp:25]: In a lot of my cases, I don't have a square grid and I don't exclusively use grid_left_bound/grid_right_bound. I had to introduce an xmin, xmax, ymin, ymax, zmin and zmax to calculate some of the limiting factors. If I set xmax, xmin for the limiter, I also have to set grid_right_bound/left_bound for the code to work, so I just put in some code that uses one or the other depending on if xmin/xmax is set.

File: src/parameters/parameters_flow_solver.h [hunk 15]
```diff
@@ -109,16 +109,14 @@ class FlowSolverParam
...
-    double grid_zmax; ///< Maximum z bound of domain for positivity_preserving_tests
-    double grid_zmin; ///< Minimum z bound of domain for positivity_preserving_tests
-
-    unsigned int number_of_grid_elements_x; ///< Number of subdivisions in x direction for positivity_preserving_tests
-    unsigned int number_of_grid_elements_y; ///< Number of subdivisions in y direction for positivity_preserving_tests
-    unsigned int number_of_grid_elements_z; ///< Number of subdivisions in z direction for positivity_preserving_tests
+    double grid_top_bound; ///< Maximum y bound of domain for a rectangle grid
+    double grid_bottom_bound; ///< Minimum y bound of domain for a rectangle grid
+    double grid_zmax; ///< Maximum z bound of domain for a rectangle grid
+    double grid_zmin; ///< Minimum z bound of domain for a rectangle grid
```
Comment (jbrillon) [inline src/parameters/parameters_flow_solver.h:115]: can we rename these to `grid_z_upper_bound` and `grid_z_lower_bound`?

File: src/flow_solver/flow_solver_cases/cube_flow_uniform_grid.cpp [hunk 16]
```diff
@@ -18,7 +18,8 @@ double CubeFlow_UniformGrid<dim,nstate>::get_adaptive_time_step(std::shared_ptr<
 {
     // compute time step based on advection speed (i.e. maximum local wave speed)
     const unsigned int number_of_degrees_of_freedom_per_state = dg->dof_handler.n_dofs()/nstate;
-    const double approximate_grid_spacing = (this->all_param.flow_solver_param.grid_right_bound-this->all_param.flow_solver_param.grid_left_bound)/pow(number_of_degrees_of_freedom_per_state,(1.0/dim));
+    double approximate_grid_spacing = 0.0;
+    approximate_grid_spacing = (this->all_param.flow_solver_param.grid_right_bound-this->all_param.flow_solver_param.grid_left_bound)/pow(number_of_degrees_of_freedom_per_state,(1.0/dim));
```
Comment (jbrillon) [inline src/flow_solver/flow_solver_cases/cube_flow_uniform_grid.cpp:22]: can we revert this change?

File: src/physics/euler.h [hunk 17]
```diff
@@ -427,6 +427,20 @@ class Euler : public PhysicsBase <dim, nstate, real>
     void boundary_farfield (
         std::array<real,nstate> &soln_bc) const;
 
+    /// Wall boundary condition
+    void boundary_do_nothing (
+        const std::array<real,nstate> &soln_int,
+        std::array<real,nstate> &soln_bc,
+        std::array<dealii::Tensor<1,dim,real>,nstate> &soln_grad_bc) const;
```
Comment (jbrillon) [inline src/physics/euler.h:434]: Can you update the doxygen comment with more details and maybe rename the function to be more specific? it seems to zero out gradients so it _does_ something

Comment (ssaishruthi98) [reply src/physics/euler.h]: Changed the name and comment to "p0 extrapolation boundary"

File: src/parameters/parameters_flow_solver.h (hunk 15 above)
Comment (ssaishruthi98) [reply src/parameters/parameters_flow_solver.h]: Changed the names to the suggested ones.

**Comment by jbrillon** [comment]
Please run ctest and post the log file here before merging :)

**Comment by ssaishruthi98** [comment]
Updated branch with latest commits from master and ran ctest. Log of failed tests is attached. All are expected failures except for `MPI_3D_EULER_NUMERICAL_ENTROPY_CONSERVATION_RRK_CPLUS` which I've logged an issue (#288) for since it is failing on master as well.
[LastTestsFailed.log](https://github.com/user-attachments/files/22647974/LastTestsFailed.log)

**Review by jbrillon — APPROVED** [review]
LGTM, thanks for making the changes!

File: src/limiter/positivity_preserving_limiter.cpp (hunk 2 above)
Comment (jbrillon) [reply src/limiter/positivity_preserving_limiter.cpp:252]: resolving for merge
